src/compiler/binder.ts

```diff
@@ -352,6 +374,272 @@ namespace ts {
             blockScopeContainer = savedBlockScopeContainer;
         }
 
+        function shouldSaveReachabilityState(n: Node): boolean {
+            return n.kind === SyntaxKind.SourceFile || n.kind === SyntaxKind.ModuleBlock || isFunctionLike(n);
+        }
+
+        function bindWithReachabilityChecks(node: Node): void {
+            let savedReachabilityState: Reachability;
+            let savedLabelStack: Reachability[];
+            let savedLabels: Map<number>;
+            let savedImplicitLabels: number[];
+            let savedHasExplicitReturn: boolean;
+
+            let saveState = shouldSaveReachabilityState(node);

```

Comment:
Perhaps just inline the check here. It's only done in this location and this is time critical code.